### PR TITLE
Complete `cosmo_array` numpy support and add `cosmo_quantity`

### LINKED ISSUE #102: There should be a cosmo_quantity analog to unyt_quantity
Many of the unyt_array functions can return a unyt_quantity in at least some circumstances. For a cosmo_array, this leads to the extra attributes being silently dropped, and then potential unexpected results or errors when trying to access those or use the resulting unyt_quantity in a calculation where the user expected a cosmo_array-like output.

### LINKED ISSUE #123: Projection/slicing returns unyt_array instead of cosmo_array
The projection and slicing functions return an `unyt_array` instead of a `cosmo_array`. Because of this, the conversion from co-moving to physical coordinates needs to be done manually. This is also a bit misleading, since this gives the impression that the resulting maps are in physical coordinates, while they are not if the coordinates in the snapshot are co-comoving.

Is there any particular reason for this? Can we just replace `unyt_array` with `cosmo_array` to fix this or does that cause other issues?

### PR DESCRIPTION
**Pull request by kyleaoman**
Unyt has a `unyt_quantity` for scalars with units attached (a subclass of `unyt_array`). It makes sense to have an analogue `cosmo_quantity` subclass of `cosmo_array`, mainly so that various numpy functions don't silently cast to `unyt_quantity`. While I'm at it am also reviewing testing of numpy function return types and cleaning up helper functions as needed.

Closes #102 
Closes #123 

So far I have new tests implemented, and the new `cosmo_quantity` class implemented. As collateral to all of this I'm implementing wrapping all of the numpy functions supported by unyt. Those are now all wrapped. I'd previously wrapped all of the numpy ufuncs so most of the logic existed already, just repackaged existing helper functions and created a couple of new ones, and now cribbed off of the unyt wrappers to make sure I don't miss any edge cases (unless they missed any!).

After a bit of tidying up and a couple of squashed bugs in unyt the test suite now passes (with the not-yet-released unyt version with my patches applied, they say hopefully released by the end of the year). Will need to pin dependencies to unyt >=3.0.4, and probably numpy >=2.0, too.

Left to do: 

- [x] Think about more test coverage.
- [x] Tidy up warnings in tests.
- [x] Consider refactoring all of the `"ca_cfs"` logic by using `cosmo_factor(None, None)` instead. That's a big job, though.
- [x] Look at a few places where `to_comoving` might be better as `convert_to_comoving` (in-place).
- [x] Look at behaviour of defaulting to comoving when mixed arguments are given, there's a kwarg now that can control this, should I use it elsewhere?
- [x] Tidy up definitions, probably move a bunch of helpers over from `objects.py` to `_array_functions.py`.
- [x] Make a PR for yt-project/unyt#549 and get it merged, then implement a wrapper for `np.take`.
- [x] Document `_array_functions.py`.
- [x] Review documentation of `objects.py`
- [x] Review/write test docstrings.
- [x] Fix `comoving == None` prints as `(Physical)`.
- [x] Look at recursion in `cosmo_array.__new__` for lists of lists of `cosmo_array`s, etc. (shouldn't recurse here because unyt doesn't, but `__new__` needs some cleanup if possible anyway).
- [x] Implement `bypass_validation`?
- [x] Look in visualisation routines to address #123 ?
- [x] Move the cosmo+unit wrapper for vis return images out of the backends and into the high-level render functions + update tests.
- [x] Review & expand narrative docs.
- [x] Review visualisation narrative docs.
- [x] Recruit some beta testers.
- [x] Check that everything is merged on unyt side and that they make a release. Will probably have to rename some stuff when they fix the fft->ftt typo.
- [x] Pin dependencies (numpy >= 2.0.0 for sure, maybe >= 2.1.0; unyt >= 3.0.4).

### CODE REVIEW

**Comment by kyleaoman** [comment]
@JBorrow just want to put this on your radar as it's basically implemented and working now (modulo notes in the post above). Gonna mostly leave it be over the xmas break, but hoping to wrap it up in January. Anyway just don't want to spring a colossal PR on you without warning - if you have any input or requests to make things go smoothly with it at this stage let me know?

**Comment by kyleaoman** [comment]
@JBorrow @robjmcgibbon oh hey the test suite passes! Still some cleanup and maybe a bit of refactoring to do but it's getting close.

**Comment by kyleaoman** [comment]
@JBorrow @robjmcgibbon I think this is mature enough to start reviewing for merge! I've got a checklist at the top mentioning the few things left to do, but those are separate from the actual code (except for looking at #123, but that's also quite peripheral to most of the changes). So I'd ask that one of you start having a look at the code changes - once I've worked through the last little things there will only be a very small amount of final review needed.

I've tried to be comprehensive with the docstrings, so suggest starting there to understand what's going on. The narrative docs (being user facing) actually won't add all that much other than to say "we support lots of numpy functions!" and give examples. If it will be some time before one of you can get to this review that's fine, but if I could request that you give me an estimate of when that'd be appreciated - just so that I can make a note of when to follow up.

File: swiftsimio/visualisation/power_spectrum.py [hunk 1]
```diff
@@ -406,7 +403,7 @@ def folded_depositions_to_power_spectrum(
                 maximally_sampled_wavenumber = np.max(
                     folded_wavenumber_centers[use_bins]
                 )
-                cutoff_wavenumber = min(
+                cutoff_wavenumber = np.min(
                     cutoff_above_wavenumber_fraction * maximally_sampled_wavenumber,
                     cutoff_above_wavenumber_fraction * cutoff_wavenumber,
                 )
```
Comment (JBorrow) [inline swiftsimio/visualisation/power_spectrum.py:409]: Can you explain to me (not necessary to change the code) why this change? Does `min` not provide the appropriate comparison in a cosmo_quantity sense?

**Comment by JBorrow** [comment]
I would _definitely_ like this to be squash merged when the time is appropriate!

File: tests/helper.py [hunk 2]
```diff
@@ -82,7 +82,7 @@ def create_single_particle_dataset(filename: str, output_name: str):
     # Create a dummy mask in order to write metadata
     data_mask = mask(filename)
     boxsize = data_mask.metadata.boxsize
-    region = [[0, b] for b in boxsize]
+    region = [[zeros_like(b), b] for b in boxsize]
```
[the remark this answers is not in the document]
Comment (kyleaoman) [reply tests/helper.py:85]: Pretty sure it still works, but might warn now, will  check.

File: swiftsimio/visualisation/power_spectrum.py (hunk 1 above)
Comment (kyleaoman) [reply swiftsimio/visualisation/power_spectrum.py:409]: Plain `min` completely bypasses the `__array_function__` method that will catch that we're dealing with numpy arrays or subclasses initially. It just compares the arguments with `__lt__`, or maybe `__gt__`, which might then get back into treating them as numpy-like. This is better for developer sanity.

File: stacked.png
[the remark this answers is not in the document]
Comment (kyleaoman) [reply stacked.png:1]: Ah yes this is persistent... they don't always get cleaned up properly after the tests run. I'll get annoyed enough to fix eventually...

File: swiftsimio/_array_functions.py [hunk 3]
```diff
@@ -0,0 +1,2179 @@
...
+
+    Parameters
+    ----------
+    input_object : :obj:`object`
+        Object to consider for promotion from unyt instance to cosmo instance.
+    """
+    if isinstance(input_object, unyt_quantity) and not isinstance(
+        input_object, objects.cosmo_quantity
+    ):
+        return input_object.view(objects.cosmo_quantity)
```
[the remark this answers is not in the document]
Comment (kyleaoman) [reply swiftsimio/_array_functions.py:215]: The idea is that there's no reason to *prevent* this working if there's a 3rd class involved somehow. I had to change some usage of `type()` in unyt to `isinstance()`, so seemed to make sense to "pay it forward"?

**Comment by kyleaoman** [comment]
> I would _definitely_ like this to be squash merged when the time is appropriate!

Indeed!

File: tests/helper.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (kyleaoman) [reply tests/helper.py:85]: Indeed, it just warns you about a missing `cosmo_factor` and states what it's assuming.

File: swiftsimio/_array_functions.py [hunk 4]
```diff
@@ -0,0 +1,2179 @@
...
+)
+
+_HANDLED_FUNCTIONS = {}
+
+# first we define helper functions to handle repetitive operations in wrapping unyt &
+# numpy functions (we will actually wrap the functions below):
+
+
+def _copy_cosmo_array_attributes(
+    from_ca: object, to_ca: object, copy_units=False
```
[the remark this answers is not in the document]
Comment (kyleaoman) [reply swiftsimio/_array_functions.py:126]: This is correct. From the docstring:
```
Copies the ``cosmo_factor``, ``comoving``, ``valid_transform`` and ``compression``
    attributes across if both the source and destination objects are
    :class:`~swiftsimio.objects.cosmo_array` instances (else returns input).
```
This just makes it easier to handle cases where the input might or might not be a `cosmo_array`. Take `numpy.meshgrid` for example. If you have one argument that's a `np.ndarray` and one that's a `cosmo_array` (there's nothing wrong with this) you'll end up in our wrapper code. This means that we can just ask to copy across attributes if present by calling the function instead of having to check each argument first.

I renamed it `_copy_cosmo_array_attributes_if_present`.

File: swiftsimio/__init__.py [hunk 5]
```diff
@@ -8,6 +8,7 @@
 import swiftsimio.metadata as metadata
 import swiftsimio.accelerated as accelerated
 import swiftsimio.objects as objects
+from swiftsimio.objects import cosmo_array, cosmo_quantity
```
[the remark this answers is not in the document]
Comment (kyleaoman) [reply swiftsimio/__init__.py:11]: Fixed. I'll set up ruff at some point?

[the remark this answers is not in the document]
Comment (kyleaoman) [reply swiftsimio/__init__.py:11]: Spoke too soon, moving it down creates a circular import issue. Just leave it for now?

File: docs/source/cosmo_array/index.rst [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+   import unyt as u
+   from swiftsimio.objects import cosmo_array, cosmo_factor
+   from swiftsimio.objects import a  # the scale factor represented by a sympy expression
+
+   my_cosmo_array = cosmo_array(
+       [1, 2, 3],
+       u.Mpc,
+       comoving=True,
+       cosmo_factor=cosmo_factor(a**1, 1.0)  # 1.0 is the current scale factor
```
[the remark this answers is not in the document]
Comment (kyleaoman) [reply docs/source/cosmo_array/index.rst]: Done (and updated the docs accordingly).

File: swiftsimio/_array_functions.py [hunk 7]
```diff
@@ -0,0 +1,2179 @@
...
+    cfs : swiftsimio.objects.cosmo_factor
+        :class:`~swiftsimio.objects.cosmo_factor`s to be multiplied.
+
+    Returns
+    -------
+    out : swiftsimio.objects.cosmo_factor
+        The product of the input :class:`~swiftsimio.objects.cosmo_factor`s.
+    """
+    if len(cfs) == 1:
+        return __binary_multiply_cosmo_factor(cfs[0])
```
[the remark this answers is not in the document]
Comment (kyleaoman) [reply swiftsimio/_array_functions.py:332]: Refactored with `reduce`.

File: swiftsimio/_array_functions.py [hunk 8]
```diff
@@ -0,0 +1,2179 @@
...
+    Returns
+    -------
+    out : swiftsimio.objects.cosmo_factor
+        The preserved :class:`~swiftsimio.objects.cosmo_factor`.
+    """
+    if len(cfs) == 1:
+        return cfs[0]
+    retval = __binary_preserve_cosmo_factor(cfs[0], cfs[1])
+    for cf in cfs[2:]:
+        retval = __binary_preserve_cosmo_factor(retval, cf)
```
[the remark this answers is not in the document]
Comment (kyleaoman) [reply swiftsimio/_array_functions.py:399]: `reduce` is much better (also for `_multiply_cosmo_factor`), refactored.

File: swiftsimio/_array_functions.py [hunk 9]
```diff
@@ -0,0 +1,2179 @@
...
+        warnings.warn(
+            f"Mixing arguments with and without cosmo_factors, continuing assuming"
+            f" provided cosmo_factor ({cf2}) for all arguments.",
+            RuntimeWarning,
+        )
+        return cf2
+    elif cf1 != cf2:
+        raise ValueError(f"Arguments have cosmo_factors that differ: {cf1} and {cf2}.")
+    elif cf1 == cf2:
+        return cf1  # or cf2, they're equal
```
[the remark this answers is not in the document]
Comment (kyleaoman) [reply swiftsimio/_array_functions.py:456]: Done.

File: swiftsimio/_array_functions.py [hunk 10]
```diff
@@ -0,0 +1,2179 @@
...
+
+
+def _power_cosmo_factor(
+    cf1: "objects.cosmo_factor",
+    cf2: "objects.cosmo_factor",
+    inputs: "Tuple[objects.cosmo_array]" = None,
```
[the remark this answers is not in the document]
Comment (kyleaoman) [reply swiftsimio/_array_functions.py:462]: Fixed (well, that argument removed). But fixed the rest of the similar instances in this file...

File: swiftsimio/_array_functions.py [hunk 11]
```diff
@@ -0,0 +1,2179 @@
...
+
+    Raises
+    ------
+    ValueError
+        If the exponent is not a dimensionless quantity, or the exponent has a
+        ``cosmo_factor`` whose scaling with scale factor is not ``1.0``.
+    """
+    if inputs is not None and power is not None:
+        raise ValueError
+    power = inputs[1] if inputs else power
```
[the remark this answers is not in the document]
Comment (kyleaoman) [reply swiftsimio/_array_functions.py:497]: This was a holdover from some earlier implementations. The `inputs` argument was never used anywhere. I've removed it.

File: docs/source/visualisation/power_spectra.rst [hunk 12]
```diff
@@ -112,7 +113,20 @@ allows you to do this easily:
...
         box_size=data.metadata.box_size,
         number_of_wavenumber_bins=128,
-        wavenumber_range=[1e-2 / unyt.Mpc, 1e2 / unyt.Mpc],
+        wavenumber_range=[
+	    cosmo_quantity(
+	        1e-2,
+	        u.Mpc**-1,
+	        comoving=True,
+	        cosmo_factor=cosmo_factor(a**-1, data.metadata.scale_factor,
+            )
```
[the remark this answers is not in the document]
Comment (kyleaoman) [reply docs/source/visualisation/power_spectra.rst:122]: Code for this is done now too - just making a note for myself that it wants a few tests for the new constructor and a revision of the narrative docs.

[the remark this answers is not in the document]
Comment (kyleaoman) [reply docs/source/visualisation/power_spectra.rst:122]: And done.

**Comment by kyleaoman** [comment]
@JBorrow @robjmcgibbon Everything that both of you have raised is now addressed. I've checked and in some cases revised the visualisation examples from the narrative docs, they now all run out of the box.

There are a small number of API changes from the user perspective - a few necessary given the changes, and a few where it seemed like this would be a good time to make them. Notably the removal of `boxsize` as an argument to a couple of visualisation functions where an object with metadata is already being passed - better to just get the box size internally in that case.

There are two minor (non-blocking, imo) outstanding issues that need changed in unyt to be rectified. These have xfailing tests in place, so we'll catch them when unyt fixes them (I already contributed one fix that's merged to their main, just depends when they release again...). The two issues are that using `np.block` will crash, and that if someone ever decided to do `U.Mpc * cosmo_array(...)` they'd get back a `unyt_array`, whereas `cosmo_array(...) * U.Mpc` gives a `cosmo_array` as you'd expect it to. The way that `__mul__` and `__rmul__` work is kind of awful, had never really dug deep into that bit of python before.

One gotcha that I have no good way of addressing is that `unyt` only traverses one level of nested containers on initialisation. This means that `cosmo_array([[x, x], [x, x]])` where `x` is a `cosmo_array` or `unyt_array` completely ignores any extra attributes of `x` and just treats it as a raw numpy array. A full-depth traversal would be a significant design change on the unyt side, not to mention being really really expensive in a lot of common use cases, so I don't think there's a strong argument to change this behaviour. It does tend to come up when you define say a mask region as `[[xmin, xmax], [ymin, ymax], [zmin, zmax]]` where all of those are `cosmo_quantity` objects... just need to get in the habit of coding around that, I guess. The docs do cover this, for all of the good that will do. Old code will usually complain with a warning here, so that's something.

Aaaaanyway, dare I say, ready to merge?

**Comment by kyleaoman** [comment]
Ah and I tested that numpy>=2.0.0 isn't enough now, so require >=2.1.0. And unyt>=3.0.4 now (just released with a bunch of fixes that I contributed throughout all of this).

**Comment by robjmcgibbon** [comment]
Fantastic, thanks Kyle, looks good to me. I will let Josh have the final say since he had more comments than me.

**Comment by kyleaoman** [comment]
I'm tinkering with the leftover non-blocking stuff and discovered a dodgy test, figured I'd sneak it in...

**Comment by kyleaoman** [comment]
I think it's ready for a release. There might be a follow-up minor release, but that would first involve wrangling some changes in unyt and getting them to release them, so rather go ahead with this. As far as I know there's nothing serious outstanding.
